python/pyspark/daemon.py

```diff
@@ -102,7 +102,7 @@ def shutdown(code):
         signal.signal(SIGTERM, SIG_DFL)
         # Send SIGHUP to notify workers of shutdown
         os.kill(0, SIGHUP)
-        sys.exit(code)
+        os._exit(code)
```

Comment:
Per the Python doc:

> _exit() should normally only be used in the child process after a fork().

Do we still trigger the cleanup handling as before? I suspect it fails to terminate its forked processes for some reasons and `os._exit` seems not guaranteeing to terminate the workers.

Can you check if they are really dead via manually checking opened file descriptors? You can check it via `lsof` if you're using Mac.

Reply:
This line `os.kill(0, SIGHUP)` kill child processes by sending SIGHUP to them.
And I prefer `os._exit` rather than `sys.exit` here, because `sys.exit` will throw `SystemExit` exception in main thread. If this `SystemExit` exception is caught by user code or there're a dead loop in user code then kill will fail.

Reply:
>  If this SystemExit exception is swallowed by user code somewhere or in finally block run into some blocking code, then the kill will fail.

`os.kill(0, SIGHUP)` is non-blocking call IIRC and seems `SystemExit` is not being swallowed in the current daemon. Why is it specific to Python 3.7?

Given my debug (IIRC), workers didn't die by `os.kill(0, SIGHUP)` at Python 3.7 specifically. My impression was that it's related to the current bug.

We should trigger the cleanup procedure per https://docs.python.org/3/library/os.html#os._exit; otherwise, it could cause another set of resource leak issues.

I still think we should identify (and possibly file a bug in Python side). `os._exit` is a way officially discouraged and we need a reason to use this instead.